New algorithm MaskBinsIf

gvardany opened the pull request:
**Description of work.**

This adds a new convenience algorithm for masking bins with criteria specified as muparser expression.

**To test:**

```
CreateSampleWorkspace(BankPixelWidth=100, NumBanks=1, OutputWorkspace='out')
MaskBinsIf(InputWorkspace='out', Criterion='s>10 && s<20 && x>1000 && x<2000', OutputWorkspace='out')
```

Open the matrix view of the workspace and check that the greyed bins are as expected. 
Play around with the criteria after having read the documentation.

<!-- Instructions for testing. -->

Fixes #23745 . <!-- and fix #xxxx or close #xxxx xor resolves #xxxx -->
<!-- alternative
*There is no associated issue.*
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
-->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

mantid-builder commented:
@gvardany do you want to request a review on this PR?
<!-- {"problem_type": "unreviewed"} -->

OwenArnold commented on Framework/Algorithms/src/MaskBinsIf.cpp:
```diff
@@ -0,0 +1,129 @@
+    const bool hasDx = outputWorkspace->hasDx(index);
+    for (auto it = spectrum.begin(); it != spectrum.end(); ++it) {
+      const auto bin = std::distance(spectrum.begin(), it);
+      y = it->counts();
```
👍 for ensuring `counts` and not just the held value.

OwenArnold commented on docs/source/algorithms/MaskBinsIf-v1.rst:
```diff
@@ -0,0 +1,35 @@
+The variables entering the criteria are reserved as follows:
+
+- y : count in a bin
+- x : the bin center
```
in arbitrary units

OwenArnold commented on docs/source/algorithms/MaskBinsIf-v1.rst:
```diff
@@ -0,0 +1,35 @@
+
+- y : count in a bin
+- x : the bin center
+- e : the error on the count
```
standard deviation

OwenArnold commented on docs/source/algorithms/MaskBinsIf-v1.rst:
```diff
@@ -0,0 +1,35 @@
+- x : the bin center
+- e : the error on the count
+- dx : the error on the bin center
+- i : the workspace index
```
Runs from 0 - number of spectrum -1. Not to be confused with the spectrum number.

As a related point, we hoped to make an effort to increase (mandate) the useage of indexes internal to Mantid, but keep (prefer) ids externally. Would it be better to change I to be the Spectrum Number. Gives us better protection from operations which affect i, such as cropping the workspace, or extracting spectra?

gvardany replied:
The spectrum number can be specified with 's' as a spectrum axis value, isn't it? If the spectrum axis is converted though, then I don't think the spectrum number makes much sense. So I thought the combination of workspace index and the spectrum axis value covers all.

OwenArnold replied:
This is a little tricky. Worth discussing I think as once published, very hard to roll-back.

In the case of the x-axis you use the actual values, not the bin indices. For spectrum axis converted to say two theta, can't we do the same?

gvardany replied:
It is the same for the spectrum axis; 's' is the value of the axis vector. But if the axis is not converted to anything, the axis value is the spectrum number, isn't it?

OwenArnold replied:
@gvardany I think so. If this is the case, can we keep `s` but drop the `i`? Does it serve any useful purpose that is needed now?

gvardany replied:
Yes, we can drop `i`; it was more like for convenience.

gvardany replied:
Actually, reading the code I realised that there might be a use case for `i`; when the spectrum axis is a text axis.

OwenArnold replied:
If you have a `text` axis, would you still ever want to apply a masking operation?

OwenArnold replied:
I really don't know the "right" answers, I'm just trying to think it through as well.

gvardany replied:
I'm not sure, one could have text labels for the spectra, but still wish to mask some of them. So in principle this can be used although I haven't seen text axes in use for a long time.

OwenArnold replied:
I'm slightly worried that the following (type) mask operations will do different things when using the `i` notation:

```
 ws = CreateSampleWorkspace(BankPixelWidth=100, NumBanks=1, OutputWorkspace='out')
 ws =  MaskBinsIf(InputWorkspace=ws, Criterion='i >10 && i<20 && x>1000 && x<2000', OutputWorkspace='out')
ws = CropWorkspace(ws, StartWorkspaceIndex=0, EndWorkspaceIndex=10)
 ws =  MaskBinsIf(InputWorkspace=ws, Criterion='i >10 && i<20 && x>1000 && x<2000', OutputWorkspace='out')
```

One option would be to refuse to process any workspace that does not have a numeric vertical axis. That way you can only provide `s`. We could add `i` at a later point if really required?

gvardany replied:
That sounds sensible I guess.

OwenArnold replied:
If you're really dead set on passing indexes in, I would be ok with that. I'm just wary of it having made the mistake myself a few times.

OwenArnold commented on Framework/Algorithms/src/MaskBinsIf.cpp:
```diff
@@ -0,0 +1,129 @@
+      e = it->countStandardDeviation();
+      dx = hasDx ? it->centerError() : 0.;
+      if (parser.Eval() != 0.) {
+        outputWorkspace->maskBin(index, bin);
```
@soininen has recently proposed a new algorithm - `MaskingNonOverlappingBins`. His policy here (my preference too) is not to clear the data that is being masked. I think the decision to `maskBin` rather than `flagMasked` should be discussed.

OwenArnold commented on docs/source/release/v3.14.0/framework.rst:
```diff
@@ -40,6 +40,7 @@ New Algorithms
 ##############
 
 - :ref:`MatchSpectra <algm-MatchSpectra>` is an algorithm that calculates factors to match all spectra to a reference spectrum.
+- :ref:`MaskBinsIf <algm-MaskBinsIf>` is an algorithm to mask bins according to criteria specified as a muparser expression.
```
I think you need to make it clear that the data is also being zero'd out as part of the masking process.

gvardany commented:
So I removed the `i`, put an error for text axis and changed `maskBins` to `flagMasked`, please have another look. I think this should be good for the moment.

OwenArnold commented on Framework/Algorithms/src/MaskBinsIf.cpp:
```diff
@@ -96,16 +94,19 @@ void MaskBinsIf::exec() {
   const auto spectrum = dynamic_cast<SpectraAxis *>(spectrumAxis);
   const bool spectrumOrNumeric = numeric || spectrum;
+  if (!spectrumOrNumeric) {
+    throw std::runtime_error(
```
👍
